draft-kazuho-httpbis-priority.md

```diff
@@ -98,6 +98,70 @@ The terms sh-token and sh-boolean are imported from
 Example HTTP requests and responses use the HTTP/2-style formatting from
 {{?RFC7540}}.
 
+# Negotiating Priorities
+
+The HTTP/2 specification defines a priority scheme in [RFC7540],
+Section 5.3, which some implementers have opted not to fully support.
+The lack of signalling about the status of the implementation has
+caused several implementations to implement heuristics to detect when
+the clients they are connected to do not support priorities as
+defined and take steps to compensate for that.
+
+The intent of this negotiation is to provide a signalling mechanism
+for each peer to communicate which, if any, priority schemes are
+supported, as well as the server's ranked preference.
+
+For both HTTP/2 and HTTP/3, either peer's SETTINGS may arrive first,
+so any negotiation must be unilateral and not rely upon receiving
+the peer's SETTINGS value.
+
+Implementations are likely to only use one prioritization scheme at
+once, and may be unable to change the scheme once established, so the
+setting MUST be sent prior to the first request if it is ever sent.
+In HTTP/3, SETTINGS may arrive after the first request even if
+they are sent first.  In order to avoid the server incorrectly
+choosing a priority scheme in HTTP/3, the client SHOULD only send
+priority information prior to receiving the server's SETTINGS if
+it only offers a single scheme.  If it offers multiple schemes,
+the client SHOULD delay sending priority information until
+after it knows what the server supports, or risk the server
+choosing a sub-optimal scheme.
+
+## The SETTINGS_PRIORITIES SETTINGS Parameter
+
+This document adds a new SETTINGS parameter to those defined by
+[RFC7540], Section 6.5.2.
+
+The new parameter name is SETTINGS_PRIORITIES, which allows both
+peers to indicate which prioritization schemes they support.
+
+A value of 0 indicates no support for priorities. If either side sends the
+parameter with a value of 0, clients SHOULD NOT send priority frames
+and servers SHOULD NOT make any assumptions based on the presence or
+lack thereof of priority frames.
+
+If the value is non-zero, then the least significant 8 bits indicates the
+peer's preferred priority scheme, the second least significant 8 bits
+indicates the peer's second choice, and so on.  This allows expressing
+support for 4 schemes in HTTP/2 and 7 in HTTP/3.  If any octet is 0,
+all more significant octets MUST also be 0.
```

Comment:
It's difficult to look ahead and know how many schemes will exist, but I'm afraid 4 (or even 7) might be low in the long run. It depends on if "schemes" will be mutually exclusive or not. This is the case for the current + H2's dep tree, but if we go with the header-based design, I can see later schemes being extensions on the current one. See for example #28 or #50. 

Put differently: If we later specify new header values, like "blocking" or "visible", in a separate draft that extends upon this one, would that be a new value? And how does that work with negotiating if a peer only supports this current draft (which bisects nicely with the new one)

Or is the "scheme" more like the "delivery method", where it indicates header-based priorities instead of specific header values/options that may be contained therein...

Reply:
This design restricts the size of the presented list but not the size of the code space. 8-bits give enough(maybe?) codepoints while allowing a reasonable amount of room for advertising support for multiple priority schemes

Reply:
I agree with that. I just wonder: if current document is value 2, and follow-up is 3 (which includes 2 and adds new values): if client sends 3 but server only has 2, what happens? If client also has to send 2 besides 3, then I can see problems down the line with only allowing 4-7 discrete values, depending on how fine-grained new definitions will be in the future. 

Hence, we allow 8 bits so we can have -many- schemes in the future. But we can only select 4-7  of those at the same time, kind of undoing/restricting the usefulness of many schemes?